### PR TITLE
Use a cleaner method to determine a build's environment #5175

### LINKED ISSUE #5175: Use a cleaner method to determine a build's environment
`-DisDevEnvironment=true` is a hackish way to do it.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #5175 

Not sure who's best to review this, though.

### CODE REVIEW

File: src/main/java/teammates/common/util/Config.java [hunk 1]
```diff
@@ -78,6 +78,13 @@ public String getAppVersion() {
         return SystemProperty.applicationVersion.get().split("\\.")[0].replace("-", ".");
     }
 
+    /**
+     * @return true if the system is running at development environment
+     */
+    public boolean isDevEnvironment() {
+        return SystemProperty.environment.value() == SystemProperty.Environment.Value.Development;
+    }
```
Comment (damithc) [inline src/main/java/teammates/common/util/Config.java]: Where is this property set? Automatically set by GAE?

Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java]: That's correct. Nothing needs to be done on the dev's side.

**Comment by damithc** [comment]
@bamboo3250 ? He is a co-lead for `t.GAE`

File: src/main/java/teammates/common/util/Config.java [hunk 2]
```diff
@@ -78,6 +78,13 @@ public String getAppVersion() {
         return SystemProperty.applicationVersion.get().split("\\.")[0].replace("-", ".");
     }
 
+    /**
+     * @return true if the system is running at development environment
+     */
+    public boolean isDevEnvironment() {
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/Config.java]: This method can be static method, right?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 3]
```diff
@@ -398,12 +398,10 @@ private static String replaceUnpredictableValuesWithPlaceholders(String content)
...
                       // date/time now e.g [Thu, 07 May 2015, 07:52 PM] or [Thu, 07 May 2015, 07:52 PM UTC]
                       .replaceAll(dateTimeNow + REGEX_DISPLAY_TIME, "\\${datetime\\.now}")
-                      // jQuery files
-                      .replace("/js/lib/jquery.min.js", "${lib.path}/jquery.min.js")
-                      .replace("https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js", "${lib.path}/jquery.min.js")
-                      // jQuery-ui files
-                      .replace("/js/lib/jquery-ui.min.js", "${lib.path}/jquery-ui.min.js")
-                      .replace("https://ajax.googleapis.com/ajax/libs/jqueryui/1.10.4/jquery-ui.min.js", "${lib.path}/jquery-ui.min.js")
+                      // jQuery js file
+                      .replace(Const.SystemParams.getjQueryFilePath(TP.isDevServer()), "${lib.path}/jquery.min.js")
```
Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Could you remind me what is the difference between DevServer and DevEnvironment? Are they the same? Can we use isDevEnvironment() here instead?

File: src/main/java/teammates/common/util/Config.java (hunk 2 above)
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java]: Yes it can, but I'm following how the two methods above this (the ones that involve `SystemProperty`) are non-static.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 3 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: No diff. `TestProperties.isDevServer()` has been there since forever. If you're aiming for uniformity, I'm the one who should make the change, i.e `Config.isDevServer()`. Is that what you mean?

Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Yes, I want to avoid "2 functions do the same thing" situation. Either one of them should be kept only.

Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: @wkurniawan07 I think we'd better remove `isDevServer()` in `TestProperties`. It's not the right way to check for the environment anyway. This url `http\://teammates-localhost.appspot.com
` would break the logic.

Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: And please rename `isDevEnvironment()` in `Config` to `isDevServer()` if possible.

Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: @bamboo3250 I've noticed that since ages ago :p but you can't remove it, it's not replaceable with `Config.isDevServer()`. I know how to fix it, though.

Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Why it isn't replaceable with `Config.isDevServer()`? Are their results the same?

Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Nope. `SystemProperty.*` (`*` can e.g be `applicationId`, `applicationVersion`, or `environment`) are meaningful only when called by a GAE-compiled class inside the application (I don't know how to better describe it, but it would include most of our production codes). Calls from any other random Java class (and this includes all our tests) will simply result in null.
You can try it yourself: try to display `SystemProperty.*.get()` from any of the test classes. I can guarantee you a null. Better yet, try and call  `Config.inst().getApplicationVersion()`: you'll get an NPE. (yes this is another bug that I've known for ages, but it's not so poisonous so I still let it)

**Comment by wkurniawan07** [comment]
@bamboo3250 you can review again. Travis is still running, but I'm confident that it will still be green (I only change the method name).

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 3 above)
Comment (bamboo3250) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I see. I tried and it did return `null` as you said. Then could you add a comment for `Config. isDevServer()` to tell not to call it from classes that are not related to GAE? Perhaps we will address this issue about these two `isDevServer()` in another PR.

**Comment by bamboo3250** [comment]
The other look fine :+1: Can merge after adding comment.

File: devdocs/settingUp.md [hunk 4]
```diff
@@ -100,9 +100,7 @@ Important: When a version is specified, please install that version instead of t
     Firefox 38.0.5 (latest release as at 7th June 2015) is supported.
    
 2. Before running the test suite, both the server and the test environment 
-   should be using the UTC time zone. The server and the test environment should 
-   also serve the CDN files (files that we off-load to servers such as Google's servers)
-   locally instead. In our case, these are files such as jQuery.min.js.
+   should be using the UTC time zone.
```
Comment (damithc) [inline devdocs/settingUp.md]: Ensure that the CDN trick is mentioned somewhere in the code. Otherwise future devs will wonder why we switch locations of the jquery files.

**Comment by wkurniawan07** [comment]
@damithc added comments for both `isDevServer` and the CDN trick.

File: src/main/java/teammates/common/util/Const.java [hunk 5]
```diff
@@ -57,6 +57,20 @@
         public static final String QUEUE_XML_PATH = "src/main/webapp/WEB-INF/queue.xml";
         public static final String DEFAULT_PROFILE_PICTURE_PATH = "/images/profile_picture_default.png";
         
+        // The library files are chosen based on the following consideration:
+        // 1. Use local files on development to enable purely offline testing
+        // 2. Use CDN files on production to reduce the load on appspot's server
+
+        public static final String getjQueryFilePath(boolean isDevServer) {
```
Comment (damithc) [inline src/main/java/teammates/common/util/Const.java]: Perhaps give it as a javadoc comment for each method? That way it will show up in the IDE when someone tries to use the method. e.g. `Returns the ... when on dev env .. but returns ... on production ... to avoid the overhead ...`
Comments in between methods are not attached to method and can detached from methods during refactoring. 
Do the same for the other three methods in `Config.java`

**Comment by damithc** [comment]
Can merge after addressing comment.
